design/20210209.certificates.k8s.io-adoption.md

```diff
@@ -0,0 +1,157 @@
+    `Issuer` types
+
+### Non-Goals
+
+- Remove the `CertificateRequest` resource in favour of
+    `CertificateSigningRequest`
+- Change the `Certificate` controllers behaviour
+- Support `CertificateSigningRequest` pre v1
+- Make changes to upstream Kubernetes to implement controllers in cert-manager
+
+
+## Reasons for Approach
+
+Below are a summary of properties of the `CertificateSigningRequest`, and
+explanations as to why other approaches to integrate the resource with
+cert-manager aren't possible (i.e. completely switching the `CertificateRequest`
+resource to `CertificateSigningRequest`:
+
+1. `CertificateSigningRequests` are cluster scoped resources. This means that
+   the concept of namespaced `Issuers` doesn't fit with this resource, and any
+   attempt to integrate it would open up the ability of privilege escalation
+   (i.e. referencing Secrets in another tenants Namespace). Switching to
+   `CertificateSigningRequests` would mean the removal of the `Issuer` resource.
+
+1. As `CertificateSigningRequests` are cluster scoped, they do not provide the
+   same kind of isolation that `CertificateRequests` provide- preventing tenants
+   reading others requests, namespace specific issuer authentication (e.g. Vault
+   Kubernetes auth), namespacing cert-manager.
+
+1. `CertificateSigningRequests` do not include a duration field. This means that
+   any attempt to "proxy", or copy `CertificateRequests`, would result in a
+   regression of this feature. A feature which many users of cert-manager rely
+   on.
+
+1. cert-manager has a large number of [external
+   issuers](https://cert-manager.io/docs/configuration/external/) created by
+   the community that operate over the `CertificateRequest` resource. No longer
+   supporting these resources would require these authors to completely re-write
+   these issuers to continue to be supported by cert-manager.
+
+Supporting both resource types, but making `CertificateRequests` the "priority"
+resource (i.e. the resource created by the `Certificates` controller), means
+that we don't have any regression in the features cert-manager offers, whilst
+maintaining the same security properties.
+
+## Proposal
+
+Instead of the concept of an `IssuerRef` for `CertificateRequests`,
+`CertificateSigningRequests` have the concept of a `SignerName`. Since
+`CertificateSigningRequests` are cluster scoped resources, the signer name can
+be directly mapped to the `ClusterIssuer` resource. `ClusterIssuers` will be
+referenced in the following format:
+
+```yaml
+  signerName: cert-manager.io/${ClusterIssuer}
+```
+
+Each `CertificateSigningRequest` controller will behave in the same way as the
+existing `CertificateRequest` resource, by getting the referenced
+`ClusterIssuer`, and attempting to sign. If the `ClusterIssuer` type is not
+managed by this controller, do nothing, else sign.
+
+Each `CertificateSigningRequest` controller will automatically set the
```

Comment:
I think we only need a single approver implementation for all CSR controllers right? (and give users an option to disable this if they want manual/their own approval)

Reply:
Wouldn't it be useful for users to have the ability to deploy multiple CA Issuers (all supported by certManager) in their cluster and direct CSR's to the appropriate CA by referencing the ClusterIssuer in the SignerName? Currently, is there a separate controller for each ClusterIssuer?

Reply:
This has been implemented within the cert-manager controller and no extra deployments are needed